### PR TITLE
fix(cli): pick the memory unit from the rounded figure, not the raw bytes

### PR DESCRIPTION
**Pull request by chinesepowered**
## What this PR does

Picks the memory unit from the figure `formatMemoryUsage` is about to print, rather than from the raw byte count, so a value that rounds up to the next boundary is no longer labelled with the unit below it.

## Why it's needed

The unit was selected by comparing the raw byte count against each boundary, but the number printed is rounded to one decimal place. Rounding can carry a value up to the next boundary while the branch that was already chosen keeps the smaller label:

| bytes | before | after |
| --- | --- | --- |
| `1048575` (one under 1 MB) | **`1024.0 KB`** | `1.0 MB` |
| `1073741823` (one under 1 GB) | **`1024.0 MB`** | `1.00 GB` |

`1024.0 KB` is a megabyte-sized number wearing a kilobyte label. This is the live memory readout, so any process sitting just under a boundary displays it.

### How to verify

```
$ npx vitest run --root packages/cli --coverage.enabled=false src/ui/utils/formatters.test.ts
      Tests  43 passed (43)

# with src/ui/utils/formatters.ts reverted and the tests kept:
      Tests  2 failed | 41 passed (43)
```

The two failures are the two rows above. The rest of the added cases are guards against over-correcting into rolling over too eagerly, and pass before and after: `0`, `512`, `1024`, `1024*1024`, `1024*1024*1024`, `1023 KB`, `1023 MB`, and `1024*1024 - 100` — the last of which rounds to `1023.9` and must therefore stay in KB.

### Evidence (Before & After)

The before/after strings are tabulated above. This value appears in the footer's memory readout; the change is to the string these pure functions return, and the tests assert it directly.

### Tested on

|     OS     | Status |
| :--------: | :----: |
|  🍏 macOS  |   ✅   |
| 🪟 Windows |   ⚠️   |
|  🐧 Linux  |   ⚠️   |

### Environment (optional)

Unit tests only.

## Risk & Scope

- Main risk or tradeoff: values within one rounding step of a boundary now print in the larger unit. That is the intent, and the guard cases pin everything else in place.
- Not validated / out of scope: `formatTokenCount` in the same file has a related oddity — `9999` prints `10.0k` while `10000` prints `10k`. I left it alone on purpose: nothing there is mislabelled, and every way of smoothing it either understates `9999` as `9k` or changes the output across the whole range above it (`999999` would move from `999k` to `1000k`). That is a formatting preference rather than a defect, so it does not belong in a fix PR.
- Breaking changes / migration notes: none.

## Linked Issues

None.

<details>
<summary>中文说明</summary>

## 本 PR 的作用

让 `formatMemoryUsage` 依据它即将打印出来的数值来选择单位，而不是依据原始字节数，这样四舍五入后进位到下一个边界的数值就不会再被标上低一级的单位。

## 为什么需要

单位是通过将原始字节数与各个边界比较来选定的，但打印出来的数字保留了一位小数。四舍五入可能把数值推进到下一个边界，而此前已经选定的分支仍然沿用较小的单位标签：

| 字节数 | 修复前 | 修复后 |
| --- | --- | --- |
| `1048575`（比 1 MB 少 1） | **`1024.0 KB`** | `1.0 MB` |
| `1073741823`（比 1 GB 少 1） | **`1024.0 MB`** | `1.00 GB` |

`1024.0 KB` 是一个兆字节量级的数字却挂着千字节的标签。这是实时的内存读数，因此任何内存占用恰好略低于边界的进程都会显示成这样。

### 如何验证

```
$ npx vitest run --root packages/cli --coverage.enabled=false src/ui/utils/formatters.test.ts
      Tests  43 passed (43)

# 回退 src/ui/utils/formatters.ts 并保留测试后：
      Tests  2 failed | 41 passed (43)
```

失败的两项正是上表中的两行。新增的其余用例都是防止过度修正（避免过早进位）的保护性用例，修复前后均通过：`0`、`512`、`1024`、`1024*1024`、`1024*1024*1024`、`1023 KB`、`1023 MB`，以及 `1024*1024 - 100`——最后这一项四舍五入后为 `1023.9`，因此必须保持在 KB。

### 证据（修复前后对比）

修复前后的字符串已在上表列出。该数值出现在底栏的内存读数中；本次改动针对的是这些纯函数返回的字符串，测试直接对其进行断言。

### 测试环境

|    操作系统    | 状态 |
| :--------: | :----: |
|  🍏 macOS  |   ✅   |
| 🪟 Windows |   ⚠️   |
|  🐧 Linux  |   ⚠️   |

### 运行环境（可选）

仅单元测试。

## 风险与影响范围

- 主要风险或权衡：与边界相差不到一个舍入步长的数值现在会以更大的单位显示。这正是本次改动的意图，而保护性用例已将其余情况固定住。
- 未验证 / 不在范围内：同一文件中的 `formatTokenCount` 存在类似的怪异之处——`9999` 显示为 `10.0k`，而 `10000` 显示为 `10k`。我有意没有改动它：那里并不存在单位标错的问题，而任何抹平这一差异的做法，要么会把 `9999` 低估为 `9k`，要么会改变其之上整个区间的输出（`999999` 会从 `999k` 变成 `1000k`）。这属于格式偏好而非缺陷，因此不应放进一个修复类 PR。
- 破坏性变更 / 迁移说明：无。

## 关联 Issue

无。

</details>

### CODE REVIEW

**Comment by chinesepowered** [comment]
Supporting precedent I found after opening this: `formatDuration` in the same file was already fixed for exactly this bug class, and the test is right above the ones I added —

```
it('should roll a sub-minute value up to "1m" when it rounds to 60s', () => {
  // 59.95s and up round to "60.0" at one decimal, which is not a valid
  // sub-minute reading; it should render as the minute it rounds to,
  // matching formatDuration(60000) === '1m'.
  expect(formatDuration(59950)).toBe('1m');
```

Same shape: a value rounded for display crosses a boundary the branch was chosen by, and the reading that comes out is not a valid one for its unit. `1024.0 KB` is the byte-scale version of `60.0s`.

That is also why I left `formatTokenCount` alone. Its `9999 -> "10.0k"` is pinned by a deliberate existing expectation, and unlike `60.0s` or `1024.0 KB` it is not an invalid reading — just a differently-formatted one.

<details><summary>中文说明</summary>

开 PR 之后我又发现了一个佐证：同一文件中的 `formatDuration` 早已针对完全相同的问题类型做过修复，其测试就紧挨在我新增的用例上方——

```
it('should roll a sub-minute value up to "1m" when it rounds to 60s', () => {
  // 59.95s 及以上在保留一位小数时会四舍五入为 "60.0"，而这并不是一个合法的
  // 亚分钟读数；它应当渲染为其四舍五入后对应的分钟数，
  // 与 formatDuration(60000) === '1m' 保持一致。
  expect(formatDuration(59950)).toBe('1m');
```

形态完全一致：用于显示而四舍五入后的数值越过了当初用来选择分支的边界，最终输出的读数对其单位而言并不合法。`1024.0 KB` 就是字节量级上的 `60.0s`。

这也正是我没有改动 `formatTokenCount` 的原因。它的 `9999 -> "10.0k"` 被一个明确写下的既有预期固定住了，而且与 `60.0s` 或 `1024.0 KB` 不同，它并不是一个非法读数，只是格式不同而已。

</details>

File: packages/cli/src/ui/utils/formatters.ts [hunk 1]
```diff
@@ -4,15 +4,28 @@
...
-  const gb = bytes / (1024 * 1024 * 1024);
-  if (bytes < 1024 * 1024) {
-    return `${(bytes / 1024).toFixed(1)} KB`;
-  }
-  if (bytes < 1024 * 1024 * 1024) {
-    return `${(bytes / (1024 * 1024)).toFixed(1)} MB`;
-  }
-  return `${gb.toFixed(2)} GB`;
+  // The unit is chosen from the rounded figure rather than the raw byte
+  // count. Rounding to one decimal can carry a value up to the next unit's
```
[the remark this answers is not in the document]
Comment (chinesepowered) [reply packages/cli/src/ui/utils/formatters.ts]: Right, and the divergence you describe is the worst outcome — the same byte count printing differently depending on which path formatted it. Rather than carry a third copy of the fix I took the consolidation option, in 8f2f4a1.

`packages/core/src/index.ts` already does `export * from './utils/formatters.js'`, so `formatMemoryUsage` was public API from core all along. Core now holds the one implementation; the CLI UI module re-exports it and `env-snapshot.ts` imports it. Net -34 lines, and the class of bug goes away rather than being fixed three times.

The case table moved to `packages/core/src/utils/formatters.test.ts` next to the implementation; the CLI test keeps a smaller set that pins the re-export, rollover included.

Verified by restoring main's core implementation with the new tests in place: the two rollover cases fail. Core and CLI consumers (`shellExecutionService`, `memoryDiagnostics`, `env-snapshot`) pass. The `AnsiOutput` failure of `getFrameController is not a function` reproduces on unmodified main and is unrelated.

<details>
<summary>中文说明</summary>

确实，您描述的这种分歧是最糟糕的结果 —— 同一个字节数会因经由哪条代码路径格式化而输出不同。与其携带第三份相同的修复，我选择了合并方案，见 8f2f4a1。

`packages/core/src/index.ts` 本来就有 `export * from './utils/formatters.js'`，因此 `formatMemoryUsage` 一直是 core 的公开 API。现在 core 持有唯一实现；CLI UI 模块转而重新导出它，`env-snapshot.ts` 直接引入它。净减 34 行，并且这一类 bug 被消除，而不是被修复三遍。

用例表已移至实现旁的 `packages/core/src/utils/formatters.test.ts`；CLI 测试保留一组较小的用例来固定重新导出行为，包含进位场景。

验证方式：保留新测试的同时把 core 实现还原为 main 版本 —— 两条进位用例失败。core 与 CLI 的消费方（`shellExecutionService`、`memoryDiagnostics`、`env-snapshot`）均通过。`AnsiOutput` 的 `getFrameController is not a function` 失败在未修改的 main 上同样复现，与本次改动无关。

</details>

**Review by wenshao — COMMENTED** [review]
⚠️ Downgraded from Approve to Comment: CI failing: Post Coverage Comment, Test (macos-latest, Node 22.x), Test (windows-latest, Node 22.x), Integration Tests (CLI, No Sandbox), review-config. Reviewed.

<details>
<summary>中文说明</summary>

⚠️ 已从批准降级为评论：CI failing: Post Coverage Comment, Test (macos-latest, Node 22.x), Test (windows-latest, Node 22.x), Integration Tests (CLI, No Sandbox), review-config。 已审查。

</details>

_— qwen3.8-max-preview via Qwen Code /review_

**Review by doudouOUC — COMMENTED** [review]
⚠️ Downgraded from Approve to Comment: CI failing: Post Coverage Comment, Test (macos-latest, Node 22.x), Test (windows-latest, Node 22.x), Integration Tests (CLI, No Sandbox), review-config. Reviewed.

<details>
<summary>中文说明</summary>

⚠️ 已从批准降级为评论：CI failing: Post Coverage Comment, Test (macos-latest, Node 22.x), Test (windows-latest, Node 22.x), Integration Tests (CLI, No Sandbox), review-config。 已审查。

</details>

_— qwen3.7-max via Qwen Code /review_

**Comment by wenshao** [comment]
## Maintainer verification — built and run locally against `c4dc5a8`

**Verdict: merge-ready.** ✅ The fix is correct, the tests actually pin it, and the behaviour change is confined to exactly the byte range that was mislabelled. Three notes below, none blocking.

Verified head `c4dc5a84487b874210a3c7315ffdf4cc223c165f` (both commits) against merge-base `3209b89f`. Two isolated `git worktree` checkouts, each with its own `node_modules`; every `@qwen-code/*` entry was confirmed to resolve **in-tree** before anything ran, so neither arm can compile the other's source.

---

### 1. The bug reproduces on three real production surfaces — and the fix clears all three

I did not stop at the pure function. Each of these drives the actual code path a user hits:

real surfaces

| surface | base renders | head renders |
| --- | --- | --- |
| TUI footer — real `ink` render of `<MemoryUsageDisplay />` | `\| 1024.0 KB` / `\| 1024.0 MB` | `\| 1.0 MB` / `\| 1.00 GB` |
| serve daemon `/workspace/env` memory cell — real builder | `1024.0 KB` / `1024.0 MB` | `1.0 MB` / `1.00 GB` |
| shell truncation notice — **real spawned subprocess**, no mocks | `…maximum captured size of 1024.0 KB…` | `…maximum captured size of 1.0 MB…` |

Surface 3 matters for the review history: it is a **core-side** consumer, so it only turns green because of the second commit. The CLI-only first commit would have left `shellExecutionService` still printing `1024.0 KB` — which is precisely the divergence the bot warned about in its inline comment. The consolidation was the right call, and I confirmed it end-to-end rather than by reading the diff.

Guard cases (`1023.9 KB`, `256.0 MB`, `186.9 MB`) render **identically on both sides** on every surface — no over-correction.

### 2. The bug is reachable by a real process, not just by a synthetic input

real process

Nothing mocked here. I grew a real Node process's genuine RSS until it landed in the disagreement window, then called each tree's own `formatMemoryUsage(process.memoryUsage().rss)`. Both arms independently settled on the **identical real RSS of 1 073 692 672 bytes** — page-aligned, 48 KB below 1 GiB:

- base → `"1024.0 MB"`
- head → `"1.00 GB"`

Worth stating explicitly, because the description's headline example (`1048575`) is the one boundary a real process *cannot* reach — Node floors around 40 MB RSS. The reachable one is MB→GB: the window `[1073689396, 1073741823]` is 52 428 bytes wide and RSS moves in 16 KB pages, so three page-aligned RSS values sit inside it. A long session sitting just under 1 GB is an ordinary state, and it displays a gigabyte-sized number wearing a megabyte label.

### 3. Blast radius: 52 479 byte values, all of them the bug

blast radius

I imported **both real implementations side by side** in one process (base taken verbatim from `3209b89f`, head from the PR source) and diffed ~41M byte counts:

- **Dense scan, every byte `0 … 1 MB+4096`** — 1 052 673 checked, **51** differ, one contiguous range `[1048525 … 1048575]`. Both edges hold: `1048524` and `1048576` are identical on both sides.
- **Dense scan, every byte `1GB-200,000 … 1GB+4096`** — 204 097 checked, **52 428** differ, one contiguous range `[1073689396 … 1073741823]`. Lower edge `1073689395` identical.
- **Strided sweep `0 … 16 GB`** — 39 860 486 checked, 122 differ, **every one inside range B**. No third range anywhere.
- **Edge inputs** — `-1`, `-1048575`, `NaN`, `±Infinity`, `0`, `1`, `1024·GB`, `MAX_SAFE_INTEGER`: **all identical**. No incidental change to degenerate inputs.
- **Invariant** "a KB/MB string never prints a number ≥ 1024": base **156 violations** in the sample, head **0**.

So the entire behaviour change *is* the bug fix — 52 479 byte values, and each one of them is a value the old code printed with the wrong unit. That is the cleanest possible answer to the risk question in the description.

### 4. The tests are not vacuous

Tests kept, source reverted to base — a test that passes both ways pins nothing:

| suite | on head | on base source | discriminating |
| --- | --- | --- | --- |
| `packages/core/src/utils/formatters.test.ts` | 13 passed | **2 failed** \| 11 passed | 2 / 13 |
| `packages/cli/src/ui/utils/formatters.test.ts` | 35 passed | **2 failed** \| 33 passed | 2 / 35 |
| real-surface harness (footer + serve env) | 10 passed | **4 failed** \| 6 passed | 4 / 10 |
| real-subprocess harness (shell notice) | 2 passed | **1 failed** \| 1 passed | 1 / 2 |

The guard cases are supposed to pass both ways — that is their job, and they do.

### 5. Downstream consumers and static gates — all clean on the head

| check | result |
| --- | --- |
| core: `shellExecutionService` + `memoryDiagnostics` + `shell` + `formatters` | **442 passed**, 0 failed |
| cli: `env-snapshot` + `doctorCommand` + `AnsiOutput` + `shellCommandProcessor` + `systemInfo` + `formatters` | **140 passed**, 0 failed |
| `prettier --check` (5 changed files) | clean |
| `eslint` (5 changed files) | clean, exit 0 |
| `tsc --noEmit -p packages/core` | exit 0 |
| `tsc --build packages/cli` (all project refs built) | exit 0, no output |

Note `AnsiOutput.test.tsx` passes here (9 tests). @chinesepowered saw `getFrameController is not a function` — that reproduces on unmodified `main` in some environments and is not from this PR, as you said; my tree just doesn't hit it.

---

## Notes (none blocking)

**1. The two "CI failing" review downgrades are a false signal.** Both @wenshao's and @doudouOUC's `/review` runs downgraded Approve → Comment citing *"CI failing: Post Coverage Comment, Test (macos-latest), Test (windows-latest), Integration Tests (CLI, No Sandbox), review-config"*. Those check runs are **`skipped`**, not failed. Across all 100 check runs on `c4dc5a8`: **25 `success`, 66 `skipped`, 9 `cancelled`, and `0` with conclusion `failure`.** The full ubuntu `Test` job, `Serve A/B`, `Real daemon E2E` and `review-pr` all passed. This looks like a `/review` bug worth fixing separately — treating `skipped` as `failing` will keep downgrading healthy PRs.

**2. The Reviewer Test Plan in the description is stale after the consolidation commit.** It still reads `Tests 43 passed (43)` for `packages/cli … formatters.test.ts`; the actual count on the head is **35**, because the case table moved to `packages/core/src/utils/formatters.test.ts` (**13**). The revert-and-rerun instruction does still reproduce exactly 2 failures, so the plan works — only the numbers are out of date, and the core test file isn't mentioned. Worth a quick description edit.

**3. Follow-up, out of scope here.** The same bug class survives in three `formatBytes` helpers under `packages/desktop`, which this PR correctly does not touch (different package, different function). Both the if-ladder and the `Math.log`-index styles produce `1024.0 KB` for `1048575` and `1024.0 MB` for `1073741823` — I ran them:
- `packages/desktop/packages/session-tools-core/src/handlers/source-test.ts:365`
- `packages/desktop/apps/electron/src/renderer/components/right-sidebar/SessionFilesSection.tsx:85`
- `packages/desktop/packages/shared/src/utils/binary-detection.ts:289` (and the sibling in `agent/browser-tool-runtime.ts`)

Not this PR's job, but now that core owns the one canonical implementation, those are natural candidates to fold in later.

**Non-issue, for the record:** GB is the top unit, so `≥ 1024 GB` still prints `1024.00 GB`. That is unchanged by this PR and there is no larger unit to roll into — not a defect.

---

The design is also the right one, not just a working one: the branch guard `Number(x.toFixed(1)) < 1024` and the printed value `x.toFixed(1)` are the *same expression*, so a chosen unit and its printed figure cannot disagree by construction. That is why the sweep finds zero violations rather than a smaller number of them.

<details>
<summary>中文说明</summary>

## 维护者验证 —— 在本地针对 `c4dc5a8` 完整构建并运行

**结论：可以合并。** ✅ 修复是正确的，测试确实钉住了这个修复，而且行为变化被精确限定在此前会标错单位的那段字节区间内。下面有三条说明，均不构成阻塞。

验证的 head 为 `c4dc5a84487b874210a3c7315ffdf4cc223c165f`（两个 commit），对比基点为 merge-base `3209b89f`。使用两个隔离的 `git worktree`，各自拥有独立的 `node_modules`；在运行任何东西之前，已确认每一个 `@qwen-code/*` 条目都解析到**本树内部**，因此两边不可能编译到对方的源码。

### 1. 该 bug 在三个真实生产界面上均可复现，修复后三处全部恢复正常

我没有止步于纯函数层面。以下每一项驱动的都是用户实际会触达的代码路径：

| 界面 | 修复前 | 修复后 |
| --- | --- | --- |
| TUI 底栏 —— 真实 `ink` 渲染 `<MemoryUsageDisplay />` | `\| 1024.0 KB` / `\| 1024.0 MB` | `\| 1.0 MB` / `\| 1.00 GB` |
| serve 守护进程 `/workspace/env` 内存单元 —— 真实构建器 | `1024.0 KB` / `1024.0 MB` | `1.0 MB` / `1.00 GB` |
| shell 输出截断提示 —— **真实 spawn 的子进程**，无任何 mock | `…maximum captured size of 1024.0 KB…` | `…maximum captured size of 1.0 MB…` |

第 3 项对本次评审历史尤为重要：它是 **core 侧**的消费方，因此它只有靠第二个 commit 才会变绿。仅有 CLI 的第一个 commit 会让 `shellExecutionService` 继续打印 `1024.0 KB` —— 这正是 bot 在行内评论中警告的那种分歧。合并为单一实现是正确的选择，我是端到端验证的，而非仅阅读 diff。

保护性用例（`1023.9 KB`、`256.0 MB`、`186.9 MB`）在每个界面上**修复前后渲染完全一致** —— 不存在过度修正。

### 2. 该 bug 对真实进程可达，并非只有构造出来的输入才会触发

此处没有任何 mock。我让一个真实的 Node 进程的实际 RSS 增长到分歧窗口内，然后调用各自树中真正的 `formatMemoryUsage(process.memoryUsage().rss)`。两边独立地落到了**完全相同的真实 RSS：1 073 692 672 字节** —— 页对齐，比 1 GiB 少 48 KB：

- 修复前 → `"1024.0 MB"`
- 修复后 → `"1.00 GB"`

这一点值得明确指出，因为描述中作为标题示例的 `1048575` 恰恰是真实进程**无法**达到的边界 —— Node 的 RSS 下限约为 40 MB。真正可达的是 MB→GB：窗口 `[1073689396, 1073741823]` 宽 52 428 字节，而 RSS 以 16 KB 页为单位变动，因此其中包含三个页对齐的 RSS 取值。长会话的内存停在略低于 1 GB 是很常见的状态，而此时显示的就是一个 GB 量级的数字挂着 MB 的标签。

### 3. 影响范围：52 479 个字节取值，且全部属于该 bug

我把**两份真实实现并排导入同一个进程**（修复前的实现逐字取自 `3209b89f`，修复后的取自本 PR 源码），对约 4100 万个字节数做了对比：

- **逐字节稠密扫描 `0 … 1 MB+4096`** —— 检查 1 052 673 个，**51** 个不同，构成单一连续区间 `[1048525 … 1048575]`。两侧边界均稳固：`1048524` 与 `1048576` 修复前后完全一致。
- **逐字节稠密扫描 `1GB-200,000 … 1GB+4096`** —— 检查 204 097 个，**52 428** 个不同，构成单一连续区间 `[1073689396 … 1073741823]`。下边界 `1073689395` 一致。
- **跨步扫描 `0 … 16 GB`** —— 检查 39 860 486 个，122 个不同，且**全部落在区间 B 内**。任何位置都不存在第三个区间。
- **边界输入** —— `-1`、`-1048575`、`NaN`、`±Infinity`、`0`、`1`、`1024·GB`、`MAX_SAFE_INTEGER`：**全部一致**。对退化输入没有任何附带改动。
- **不变量**「KB/MB 字符串中的数字永远不会 ≥ 1024」：修复前样本中 **156 处违反**，修复后 **0 处**。

也就是说，整个行为变化*就是*这次 bug 修复本身 —— 52 479 个字节取值，每一个都是旧代码会用错误单位打印的值。这是对描述中风险问题最干净的回答。

### 4. 测试并非空转

保留测试、把源码回退到修复前 —— 两边都通过的测试钉不住任何东西：

| 测试套件 | 修复后 | 对修复前源码 | 有判别力 |
| --- | --- | --- | --- |
| `packages/core/src/utils/formatters.test.ts` | 13 通过 | **2 失败** \| 11 通过 | 2 / 13 |
| `packages/cli/src/ui/utils/formatters.test.ts` | 35 通过 | **2 失败** \| 33 通过 | 2 / 35 |
| 真实界面 harness（底栏 + serve env） | 10 通过 | **4 失败** \| 6 通过 | 4 / 10 |
| 真实子进程 harness（shell 提示） | 2 通过 | **1 失败** \| 1 通过 | 1 / 2 |

保护性用例本就应当两边都通过 —— 那是它们的职责，它们做到了。

### 5. 下游消费方与静态检查 —— 在 head 上全部干净

| 检查项 | 结果 |
| --- | --- |
| core：`shellExecutionService` + `memoryDiagnostics` + `shell` + `formatters` | **442 通过**，0 失败 |
| cli：`env-snapshot` + `doctorCommand` + `AnsiOutput` + `shellCommandProcessor` + `systemInfo` + `formatters` | **140 通过**，0 失败 |
| `prettier --check`（5 个改动文件） | 干净 |
| `eslint`（5 个改动文件） | 干净，exit 0 |
| `tsc --noEmit -p packages/core` | exit 0 |
| `tsc --build packages/cli`（所有 project ref 均已构建） | exit 0，无输出 |

补充：`AnsiOutput.test.tsx` 在我这里是通过的（9 个测试）。@chinesepowered 遇到的 `getFrameController is not a function` 在某些环境下于未修改的 `main` 上同样复现，正如你所说与本 PR 无关；只是我的环境没有触发它。

---

## 说明（均不阻塞）

**1. 两条「CI failing」的评审降级是误报。** @wenshao 与 @doudouOUC 的 `/review` 都以 *"CI failing: Post Coverage Comment, Test (macos-latest), Test (windows-latest), Integration Tests (CLI, No Sandbox), review-config"* 为由将 Approve 降级为 Comment。这些 check run 的结论是 **`skipped`**，并非失败。在 `c4dc5a8` 的全部 100 个 check run 中：**25 个 `success`、66 个 `skipped`、9 个 `cancelled`，结论为 `failure` 的有 `0` 个。** 完整的 ubuntu `Test` job、`Serve A/B`、`Real daemon E2E` 和 `review-pr` 全部通过。这看起来是 `/review` 自身的一个缺陷，值得单独修复 —— 把 `skipped` 当作 `failing` 会持续误伤健康的 PR。

**2. 描述中的 Reviewer Test Plan 在合并实现的那个 commit 之后已经过时。** 其中仍写着 `packages/cli … formatters.test.ts` 为 `Tests 43 passed (43)`；而 head 上的实际数量是 **35**，因为用例表已经移到了 `packages/core/src/utils/formatters.test.ts`（**13**）。回退并重跑的验证步骤确实仍会复现恰好 2 个失败，所以方案本身有效 —— 只是数字过时了，且没有提到 core 的测试文件。建议顺手更新一下描述。

**3. 后续跟进，不属于本 PR 范围。** 同类 bug 仍存在于 `packages/desktop` 下的三个 `formatBytes` 辅助函数中，本 PR 正确地没有触碰它们（不同 package、不同函数名）。if 分支写法与 `Math.log` 取下标写法都会把 `1048575` 打印成 `1024.0 KB`、把 `1073741823` 打印成 `1024.0 MB` —— 我实际运行验证过：
- `packages/desktop/packages/session-tools-core/src/handlers/source-test.ts:365`
- `packages/desktop/apps/electron/src/renderer/components/right-sidebar/SessionFilesSection.tsx:85`
- `packages/desktop/packages/shared/src/utils/binary-detection.ts:289`（以及 `agent/browser-tool-runtime.ts` 中的同类实现）

这不是本 PR 的职责，但既然 core 现在拥有了唯一的规范实现，这几处是后续自然的合并候选。

**另附一条非问题说明：** GB 是最高单位，因此 `≥ 1024 GB` 仍会打印为 `1024.00 GB`。这一点本 PR 未作改动，而且也不存在更高一级的单位可供进位 —— 并非缺陷。

---

这个设计不仅仅是「能用」，而且是正确的：分支判断条件 `Number(x.toFixed(1)) < 1024` 与实际打印的值 `x.toFixed(1)` 是*同一个表达式*，因此所选单位与其打印出的数字在构造上就不可能不一致。这也正是扫描结果中违反数为 0 而非某个较小数值的原因。

</details>

**Review by wenshao — APPROVED** [review]

**Comment by wenshao** [comment]
@qwen-code /triage

**Comment by yiliang114** [comment]
Review posted successfully on [PR #7871](https://github.com/QwenLM/qwen-code/pull/7871#pullrequestreview-4795814944).

**Summary**: Reviewed the `formatMemoryUsage` unit rollover fix across 5 files (+86/-37). The fix is correct — picking the unit from the rounded figure ensures the displayed value and its unit can never disagree. The consolidation of 3 duplicate copies into one core implementation is the right design. No P0/P1 findings. Posted 2× P2 (suggesting exact `toFixed(1)` boundary tests following the `formatDuration` precedent) and 3× P3 (minor quality/naming). The PR already has wenshao's approval.

---
✅ completed

**Review by yiliang114 — COMMENTED** [review]
## Code Review — `formatMemoryUsage` unit rollover fix

**Verdict:** Clean, well-scoped fix. No P0/P1 findings. 2× P2 (test suggestions), 3× P3 (minor).

The consolidation into a single core implementation is the right call — the guard expression `Number(x.toFixed(1)) < 1024` and the printed value `x.toFixed(1)` are the same expression, so the chosen unit and its printed figure cannot disagree by construction. That wenshao's dense byte-sweep found zero violations (not "fewer" violations) confirms this.

---

### 🟡 P2 — Test: Missing exact `toFixed(1)` rounding boundary tests

**File:** `packages/core/src/utils/formatters.test.ts`

The existing tests cover `1048575` (rolls to `1.0 MB`) and `1048476` / `1024*1024 - 100` (stays `1023.9 KB`), which proves the behavior on both sides of the boundary. But the exact `toFixed(1)` transition point — where `1023.95` rounds up to `1024.0` — is not tested.

The same codebase's `formatDuration` tests this exact pattern (`59949ms` → `59.9s` stays, `59950ms` → `1m` rolls). Adding a boundary-pair for `formatMemoryUsage` would catch rounding-logic regressions with higher precision:

```ts
it.each([
  [1024 * 1024 - 52, '1023.9 KB'],
  [1024 * 1024 - 51, '1.0 MB'],
])('toFixed boundary at KB→MB: %d', (bytes, expected) => {
  expect(formatMemoryUsage(bytes)).toBe(expected);
});
```

Same gap exists for MB→GB (the exact `toFixed(1)` transition is untested).

### 🟢 P3 — Quality: Redundant division in KB/MB branches

**File:** `packages/core/src/utils/formatters.ts`

`bytes / BYTES_PER_KB` is computed twice in the KB branch — once inside `roundedTo()` and again in the template literal. Same for MB. Could hoist into a local:

```ts
const kb = bytes / BYTES_PER_KB;
if (roundedTo(kb, 1) < 1024) return `${kb.toFixed(1)} KB`;
```

Negligible perf impact (display function), but avoids the risk of one occurrence being changed without the other.

### 🟢 P3 — Test: Group naming slightly misleading

`[1024 * 1024, '1.0 MB']` and `[1024 * 1024 * 1024, '1.00 GB']` are in the "keeps in own unit" group, but they're actually boundary cases where the rounded figure reaches exactly 1024.0 and triggers a unit switch.

### 🟢 P3 — Test: No invalid-input tests

No tests for negative, `NaN`, or `Infinity` inputs. Low priority since `process.memoryUsage()` always returns valid non-negative values, but the function signature accepts any `number`.

---

*Reviewed via Qwen Code pr-review skill.*

**Comment by wenshao** [comment]
@qwen-code /triage
